common/remote/hooks/git/gitfile.cpp

```diff
@@ -319,15 +319,33 @@ static IFile *createGitFile(const char *gitFileName)
     splitGitFileName(fname, gitDirectory, revision, relDir);
     if (relDir.isEmpty())
         return new GitRepositoryFile(fname, 0, true, true);  // Special case the root - ugly but apparently necessary
-    Owned<IDirectoryIterator> dir = createGitRepositoryDirectoryIterator(fname, NULL, false, true);
-    if (dir->first())
-    {
-        Linked<IFile> file = &dir->query();
-        assertex(!dir->next());
-        return file.getClear();
+    VStringBuffer gitcmd("git --git-dir=%s cat-file --batch-check", gitDirectory.get());
+    VStringBuffer lookup("%s:%s", revision.length() ? revision.get() : "HEAD", relDir.length() ? relDir.get() : "");
+    Owned<IPipeProcess> pipe = createPipeProcess();
+    if (pipe->run(nullptr, gitcmd, ".", true, true, false, 0))
+    {
+        pipe->write(lookup.length(), lookup.str());
+        pipe->closeInput();
+        StringBuffer result;
+        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+        readSimpleStream(result, *pipeReader);
+        // Should either be of the form "blobid type size" or "<ref> missing"
+        char *sizeptr = (char *) strrchr(result, ' ');
+        if (sizeptr)
+        {
+            *sizeptr++ = 0;
```

Comment:
minor: Could avoid this (and the cast above) if you used strncmp(typeptr, "tree ", 5) instead.

Reply:
details....